### PR TITLE
net: pkt: Add possibility to reserve space for link layer header

### PR DESCRIPTION
**Pull request by jukkar**
Allow network device driver to configure the system so that when a network packet is sent to it, the link layer header is stored just before the L2 payload in the same packet.
 
Currently the link layer header is stored in a separate net_buf that is linked in front of the L2 payload.

This option can typically save one net_buf when sending a network packet. Note that if you are using variable data size buffers (CONFIG_NET_BUF_VARIABLE_DATA_SIZE) then this embedding is not needed because one can allocate just the right size network buffers and not waste any memory.

Add support to Ethernet L2 to select at runtime whether to have a separate link layer header, or to embed it to first net_buf.

### CODE REVIEW

File: include/zephyr/net/ethernet.h [hunk 1]
```diff
@@ -196,6 +196,14 @@ enum ethernet_hw_caps {
 
 	/** TX-Injection supported */
 	ETHERNET_TXINJECTION_MODE	= BIT(20),
+
+	/** Hint to upper layers that this Ethernet driver would like to to have
+	 * link layer header embedded with L3 data to first packet.
+	 * By default the link layer header is placed to a separate net_buf
+	 * before the L2 payload. The driver would always be prepared to handle
+	 * the default case when link layer net_buf is a separate one.
```
Comment (pdgendt) [inline include/zephyr/net/ethernet.h:204]:
```suggestion
	/** Hint to upper layers that this Ethernet driver would like to have
	 * a link layer header embedded with L3 data in the first packet.
	 * By default the link layer header is placed in a separate net_buf
	 * before the L2 payload. The driver should always be prepared to
	 * handle the default case when link layer net_buf is a separate one.
```

File: include/zephyr/net/net_if.h [hunk 2]
```diff
@@ -622,6 +622,18 @@ struct net_if_dev {
...
+	/** The link layer header size. If set to 0 (default), a separate net_buf
+	 * containing the link layer header is placed in the chain of net_bufs when
+	 * allocating the network buffers for sending.
+	 * If this is > 0, and the corresponding L2 supports this, then a separate
+	 * headroom space is allocated to the first net_buf right before the L3
+	 * header. If the L2 does not support this, then this value is ignored.
+	 * Note that this cannot be larger than CONFIG_NET_BUF_DATA_SIZE if fixed
+	 * size net_bufs are used.
+	 * The value is not used when network packet is received.
+	 */
```
Comment (pdgendt) [inline include/zephyr/net/net_if.h:635]:
```suggestion
	 * If this is > 0, and the corresponding L2 supports this, Link Layer
	 * headroom space is allocated in the first net_buf before the L3
	 * header. If the L2 does not support this, this value is ignored.
	 * Note that this cannot be larger than CONFIG_NET_BUF_DATA_SIZE if fixed
	 * size net_bufs are used.
	 * The value is not used when a network packet is received.
	 */
```

File: include/zephyr/net/net_if.h [hunk 3]
```diff
@@ -2934,6 +2946,27 @@ int net_if_set_name(struct net_if *iface, const char *buf);
  */
 int net_if_get_by_name(const char *name);
 
+/**
+ * @brief Get the size of the embedded link layer header.
+ *        If the link layer header is not embedded to first
+ *        net_buf but is a separate net_buf, then this function
+ *        will return 0.
```
Comment (pdgendt) [inline include/zephyr/net/net_if.h:2953]:
```suggestion
 * @brief Get the size of the embedded link layer header.
 *        If the link layer header is not reserved in the first
 *        net_buf and needs a separate net_buf, then this function
 *        will return 0.
```

File: include/zephyr/net/net_pkt.h [hunk 4]
```diff
@@ -194,6 +194,11 @@ struct net_pkt {
 	uint8_t l2_processed : 1; /* Set to 1 if this packet has already been
 				   * processed by the L2
 				   */
+	uint8_t l2_header_embedded : 1; /* Set to 1 if the link layer headers
+					 * are embedded to the first packet
+					 * together with the L2 payload.
+					 * This is only used in packet send.
+					 */
```
Comment (pdgendt) [inline include/zephyr/net/net_pkt.h:201]:
```suggestion
	uint8_t l2_header_embedded : 1; /* Set to 1 if the link layer headers
					 * are embedded in the first packet
					 * together with the L2 payload.
					 * This is only used in packet send.
					 */
```

**Review by pdgendt — DISMISSED** [review]
Some comment improvements, otherwise LGTM

**Comment by jukkar** [comment]
* Updated according to comments

File: include/zephyr/net/ethernet.h [hunk 5]
```diff
@@ -196,6 +196,15 @@ enum ethernet_hw_caps {
 
 	/** TX-Injection supported */
 	ETHERNET_TXINJECTION_MODE	= BIT(20),
+
+	/**
+	 * Hint to upper layers that this Ethernet driver would like to to have
```
Comment (pdgendt) [inline include/zephyr/net/ethernet.h:201]:
```suggestion
	 * Hint to upper layers that this Ethernet driver would like to have
```

File: subsys/net/l2/ethernet/ethernet.c [hunk 6]
```diff
@@ -507,22 +507,48 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,
...
+		net_eth_is_vlan_enabled(ctx, iface) &&
+		net_pkt_vlan_tag(pkt) != NET_VLAN_TAG_UNSPEC;
+	if (is_vlan) {
+		orig_iface = net_eth_get_vlan_iface(iface, net_pkt_vlan_tag(pkt));
 	}
 
-	if (IS_ENABLED(CONFIG_NET_VLAN) &&
-	    net_eth_is_vlan_enabled(ctx, iface) &&
-	    net_pkt_vlan_tag(pkt) != NET_VLAN_TAG_UNSPEC) {
+	reserve_ll_header = net_if_reserve_ll_header_size(orig_iface);
```
Comment (pdgendt) [inline subsys/net/l2/ethernet/ethernet.c:524]: Should we check length here?

```c
if (reserve_ll_header > CONFIG_NET_BUF_DATA_SIZE) {
	return NULL;
}
```

Comment (jukkar) [reply subsys/net/l2/ethernet/ethernet.c:524]: There is already an assert for that in net_if.c `net_if_reserve_ll_header_size()` it should be enough.

File: subsys/net/ip/net_pkt.c [hunk 7]
```diff
@@ -1177,23 +1211,37 @@ int net_pkt_alloc_buffer(struct net_pkt *pkt,
...
 	net_pkt_append_buffer(pkt, buf);
 
+	/* Hide the link layer header for now. The space is used when
+	 * link layer header needs to be written to the packet by L2 send.
+	 */
+	if (reserve_ll_header > 0U) {
+		NET_DBG("Reserving %zd bytes for L2 header", reserve_ll_header);
+
+		net_buf_add(pkt->buffer, reserve_ll_header);
+		net_buf_pull(pkt->buffer, reserve_ll_header);
```
Comment (rlubos) [inline subsys/net/ip/net_pkt.c:1240]: Wouldn't it be more fit to use dedicated `net_buf_reserve()` here? The final outcome should be the same, data pointer moved by the reserved space length.

**Review by rlubos — COMMENTED** [review]
One suggestion, looks good otherwise

File: subsys/net/l2/ethernet/ethernet.c (hunk 6 above)
Comment (pdgendt) [reply subsys/net/l2/ethernet/ethernet.c:524]: Better safe than sorry?

Comment (jukkar) [reply subsys/net/l2/ethernet/ethernet.c:524]: The header size is suppose to be set by the Ethernet or VLAN code and with a proper value, so a component we are trusting. If someone else writes some garbage to it, we do assert. Isn't this pretty much the way of working in zephyr?

File: subsys/net/ip/net_pkt.c (hunk 7 above)
Comment (jukkar) [reply subsys/net/ip/net_pkt.c:1240]: Sure, we can do it like that too.

File: include/zephyr/net/net_pkt.h [hunk 8]
```diff
@@ -483,25 +492,6 @@ static inline void net_pkt_set_ip_ecn(struct net_pkt *pkt, uint8_t ecn)
 #endif
 }
 
-static inline uint8_t net_pkt_tcp_1st_msg(struct net_pkt *pkt)
```
Comment (pdgendt) [inline include/zephyr/net/net_pkt.h:486]: This is a public API and should probably follow deprecation guidelines.

Comment (jukkar) [reply include/zephyr/net/net_pkt.h:486]: There are no one in the stack that calls it. Originally it was used by the TCP for its internal housekeeping but for some reason it was not removed when the calls in tcp.c were removed. I think we can remove it safely at this point.

**Review by rlubos — DISMISSED** [review]

**Review by krish2718 — DISMISSED** [review]

File: include/zephyr/net/ethernet.h [hunk 9]
```diff
@@ -196,6 +196,15 @@ enum ethernet_hw_caps {
 
 	/** TX-Injection supported */
 	ETHERNET_TXINJECTION_MODE	= BIT(20),
+
+	/**
+	 * Hint to upper layers that this Ethernet driver would like to have
```
Comment (tbursztyka) [inline include/zephyr/net/ethernet.h:201]: This is not really a hardware capability.

File: include/zephyr/net/net_pkt.h (hunk 8 above)
Comment (tbursztyka) [reply include/zephyr/net/net_pkt.h:486]: This should anyway go into another PR, it's unrelated to ethernet afaik.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
What about having an ethernet internal net_buf pool just for headers? Then allocating a net_buf just for the header as it is would not eat up too much memory. Could be made optional as well, etc...
It would avoid having to change net_pkt etc..., moreover this seems to be pretty much an ethernet centric issue.

**Comment by jukkar** [comment]
> What about having an ethernet internal net_buf pool just for headers? Then allocating a net_buf just for the header as it is would not eat up too much memory. Could be made optional as well, etc...

That is one option of course. But then it would be specific to Ethernet, what is now in the PR can be used by any technology.

> It would avoid having to change net_pkt etc..., moreover this seems to be pretty much an ethernet centric issue.

That was not the main point, it was just easy to implement this for Ethernet. The feature is totally optional and the old way is the default one.

File: include/zephyr/net/ethernet.h (hunk 9 above)
Comment (jukkar) [reply include/zephyr/net/ethernet.h:201]: True, but it is a driver capability, we have already several of them in the capability list. There does not seem to be easy way the driver to say that it could support this.

**Comment by jukkar** [comment]
The TCP commit was there as I noticed a bit field in net_pkt that was not used any more. I can certainly send a PR for that if needed.

File: include/zephyr/net/net_pkt.h (hunk 8 above)
Comment (jukkar) [reply include/zephyr/net/net_pkt.h:486]: The TCP obsolete bit removal commit is now in #71432

**Comment by tbursztyka** [comment]
> That is one option of course. But then it would be specific to Ethernet, what is now in the PR can be used by any technology.

But is there actually a need for some tech other than ethernet?  Does not seem so, unless I am missing something

**Comment by jukkar** [comment]
> But is there actually a need for some tech other than ethernet?

That is correct, atm there are none.

I used this route (implementing having true headroom in first net_buf) so that we could avoid the extra net_buf just for the L2 header. I did not see any easy way to implement this without having this functionality implemented in net_buf. It seems to work just fine. If the change feels too controversial, we could add a Kconfig option for it that would be turned off by default, and user could select it if really needed.

**Comment by jukkar** [comment]
Changes:
* Upgraded to latest main
* Added separate Kconfig option that depends on fixed data size and Ethernet to be able to use this

**Review by rlubos — DISMISSED** [review]

File: subsys/net/ip/net_pkt.c [hunk 10]
```diff
@@ -315,10 +315,10 @@ void net_pkt_print_frags(struct net_pkt *pkt)
 	while (frag) {
 		total += frag->len;
 
-		frag_size = frag->size;
```
Comment (tbursztyka) [inline subsys/net/ip/net_pkt.c:318]: That's also a valuable commit that could go on its own PR I believe

**Comment by tbursztyka** [comment]
> > But is there actually a need for some tech other than ethernet?
> 
> That is correct, atm there are none.
> 
> I used this route (implementing having true headroom in first net_buf) so that we could avoid the extra net_buf just for the L2 header. I did not see any easy way to implement this without having this functionality implemented in net_buf.

I believe it's actually much easier than what you did with net_pkt. There would be no generic net_pkt API changes anywhere, just a dedicated pool in susbsys/net/l2/ethernet.c and one ethernet public function to allocate from it. And no existing logic changes besides using that allocator at the right place (in ethernet.c and of course in the drivers when relevant) and that's it. net_buf_unref() would do the job to deallocate to the right pool anyway.

Fact is that we use to have some sort of headroom/tailroom back then in net_pkt (if I am remembering well), and we removed it for the same reason that it was useful only in L2, and only in ethernet.

**Comment by krish2718** [comment]
> > > But is there actually a need for some tech other than ethernet?
> > 
> > 
> > That is correct, atm there are none.
> > I used this route (implementing having true headroom in first net_buf) so that we could avoid the extra net_buf just for the L2 header. I did not see any easy way to implement this without having this functionality implemented in net_buf.
> 
> I believe it's actually much easier than what you did with net_pkt. There would be no generic net_pkt API changes anywhere, just a dedicated pool in susbsys/net/l2/ethernet.c and one ethernet public function to allocate from it. And no existing logic changes besides using that allocator at the right place (in ethernet.c and of course in the drivers when relevant) and that's it. net_buf_unref() would do the job to deallocate to the right pool anyway.
> 
> Fact is that we use to have some sort of headroom/tailroom back then in net_pkt (if I am remembering well), and we removed it for the same reason that it was useful only in L2, and only in ethernet.

But with this we would end up using an extra buf (though it fixes the wastage issue with current design), no? The idea if to avoid an extra `net_pkt_read` to flatten the buffers, and use zero-copy approach for most use cases (network buffer sizes configure appropriatley).

File: subsys/net/ip/net_pkt.c (hunk 10 above)
Comment (jukkar) [reply subsys/net/ip/net_pkt.c:318]: Lets keep this commit here for time being. I can move it later if this PR gets delayed.

**Comment by jukkar** [comment]
@tbursztyka please take a look to the changes again.
The reason why I am proposing this change is to avoid having a separate net_buf for L2 in the first place. The driver in question does not support scatter/gather type interface so it needs to linearize the net_buf always. This helps to solve this issue. Also a nice benefit is to have zero copy when sending the data.
The code indeed needs to touch net_pkt internals but no API changes were needed. Also the changes are used only in Ethernet so no other type driver can try to use this by mistake. The changes also require config option to be set and also the driver needs sets a runtime flag to make this happen. Chances are quite small that this change would cause problems in other part of the system. We could mark the relevant Kconfig option as experimental to help describe the nature of the option.

File: subsys/net/ip/Kconfig [hunk 11]
```diff
@@ -700,6 +700,22 @@ config NET_PKT_BUF_USER_DATA_SIZE
 	help
 	  User data size used in rx and tx network buffers.
 
+config NET_PKT_ALLOW_HEADROOM_ALLOCATION
+	bool "Allow headroom allocation for L2 header"
+	depends on NET_BUF_FIXED_DATA_SIZE
+	depends on NET_L2_ETHERNET
```
Comment (pdgendt) [inline subsys/net/ip/Kconfig:706]: I would remove this (hard) dependency, it is currently only used for L2 ethernet, but it could be used out-of-tree.

Comment (jukkar) [reply subsys/net/ip/Kconfig:706]: I only added that to emphasize that currently this is only for Ethernet but probably it is better to not hard code this. I will remove it in next version.

**Comment by tbursztyka** [comment]
I had started a reply, and forgot to press "comment", sorry for the delay.

I do understand the requirement, that's not the issue. The problem is that it brings back an implementation that we removed because it was not well separating l2 with upper layers (i.e.: bringing an l2 attribute into net_pkt structure) 

So here is how it could be done: 
- have a generic struct net_l2 net_pkt allocator function pointer.
- in net_pkt side: if such allocator exists for the given net_if's L2, use it (after it estimated the size in net_pkt_alloc_buffer)
- also in net_pkt, perhaps rename pkt_alloc_buffer and make it public so l2 can use it directly (it's a very short function, so it could also be fully re-implemented in l2's allocator. either way)
- in L2's allocator, get the alloc_len, add the headroom you want, call pkt_alloc_buffer - or reimplement - for allocating the net_buf, then use net_buf's headroom feature to tell about the headroom (net_buf_reserve I believe?). 

It's a quick idea, I did not look in the details what could be the pitfalls. But:
- it would not require any net_pkt structure attribute for such l2 attribute
- as soon as an L2 has an allocator, it would be generalized for all devices that belong to it. So no need of a specific capability to expose. We could save a net_buf for all ethernet devices then.

Would that make any sense?

File: subsys/net/ip/net_pkt.c (hunk 10 above)
Comment (jukkar) [reply subsys/net/ip/net_pkt.c:318]: I sent a separate PR for this fix #72148

**Comment by jukkar** [comment]
> as soon as an L2 has an allocator, it would be generalized for all devices that belong to it. So no need of a specific capability to expose. We could save a net_buf for all ethernet devices then.

I have been looking this and I think this L2 needs to be device specific and needs to be able to select at runtime. We can have multiple network interfaces belonging to same L2, for example WiFi and Ethernet could be existing at the same time and they share the same L2. The separate L2 header vs embedding it to the first buf could confuse the driver if it is not prepared for it. With this in mind, the driver should be able to tell if it could use L2 header embedding.

**Comment by jukkar** [comment]
@tbursztyka please take a look, was this what you had in mind?

**Comment by jukkar** [comment]
> Where would a confusion in the driver come from?

I faintly recall seeing some weird handling of the net_pkt in some driver, but did not spot anything suspicious in the upstream drivers when I just looked. Anyway, perhaps it is not really a problem after all.

I can then remove the separate Ethernet flag as it seems it is not really needed after all. Just wondering should we still make this optional feature or always enabled for Ethernet. WDYT?

**Comment by jukkar** [comment]
* Add a Kconfig option to Ethernet to enable reserving header vs. allocating separate net_buf for it (default)
* Removed separate Ethernet capability flag, this means that all the Ethernet/Wifi interfaces in the system will use the same way to allocate L2 header

Ping @tbursztyka, please take a look.

File: include/zephyr/net/net_pkt.h [hunk 12]
```diff
@@ -1882,6 +1894,31 @@ int net_pkt_alloc_buffer(struct net_pkt *pkt,
...
+ * @param size    The size of buffer being requested.
+ * @param reserve The L2 header size to reserve. This can be 0, in which case
+ *                the L2 header is placed into a separate net_buf.
+ * @param proto   The IP protocol type (can be 0 for none).
+ * @param timeout Maximum time to wait for an allocation.
+ *
+ * @return 0 on success, negative errno code otherwise.
+ */
+#if !defined(NET_PKT_DEBUG_ENABLED)
+int net_pkt_alloc_buffer_with_reserve(struct net_pkt *pkt,
```
Comment (tbursztyka) [inline include/zephyr/net/net_pkt.h:1915]: Afaik, you don't need a dedicated net_pkt function just to be able to call net_buf_reserve() because ... (see my comment in ethernet.c)

File: subsys/net/l2/ethernet/ethernet.c [hunk 13]
```diff
@@ -774,8 +836,28 @@ enum net_l2_flags ethernet_flags(struct net_if *iface)
...
+{
+	size_t reserve_ll_header;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_NET_L2_ETHERNET_RESERVE_HEADER)) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+
+	reserve_ll_header = get_reserve_ll_header_size(iface);
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/ethernet.c:851]: your dedicated allocator knows the reserver size obviously, so after allocating the net_pkt, it could run net_buf_reserve() by itself.

Way less changes needed in net_pkt

File: subsys/net/l2/virtual/virtual.c [hunk 14]
```diff
@@ -164,8 +164,35 @@ enum net_l2_flags virtual_flags(struct net_if *iface)
...
+	}
+
+	ret = api->get_config(iface, VIRTUAL_INTERFACE_CONFIG_TYPE_L2_RESERVE,
+			      &config);
+	if (ret) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+
+	ret = net_pkt_alloc_buffer_with_reserve(pkt, size, config.l2_reserve,
```
Comment (tbursztyka) [inline subsys/net/l2/virtual/virtual.c:188]: same here

**Review by tbursztyka — DISMISSED** [review]
Unless I am missing something, we can avoid exposing any reserve parameter to net_pkt

File: subsys/net/l2/ethernet/ethernet.c (hunk 13 above)
Comment (jukkar) [reply subsys/net/l2/ethernet/ethernet.c:851]: Problem is that now every allocator needs to do the magic things like calling `net_buf_reserve()`, init cursor etc. So we are duplicating the functionality to multiple places which increases chances that something goes wrong. For Ethernet, the reserve setting would need to be done in two places, in ethernet.c and in virtual.c for VLAN. This might look like a small thing, but if some related thing need to be fixed, one would need to remember to fix it in multiple places. And this would be difficult to do for out of tree code, which does seem quite unlikely atm but still possible.
To me it really looks like it would be better if we concentrate the reserve allocation / related initialization activities to net_pkt.c.
@tbursztyka WDYT?

Comment (tbursztyka) [reply subsys/net/l2/ethernet/ethernet.c:851]: Looks like we are now with a list of pros/cons on a few things:

Generalizing the reserve into net_pkt indeed has the advantage to centralize all about allocation.
However, it unilaterally bloats net_pkt even for systems that will never need it. (note also this fix is for Ethernet only so far, question is: do we foresee any other l2 layer that could take advantage of this?)

On one hand, net_pkt_*_reserve is after all like net_buf_reserve, it brings a generic concept about reservation without the l2 one which is far better that earlier proposal.
On the other hand, it could be fully hidden from net_pkt public API as I proposed.

One fact to add: if we generalize it but only ethernet will use it, it will be more annoying to specialize it back - if ever required one day. The reverse is not an issue.

I am withdrawing my lock on this PR. I let other reviewers (@pdgendt @rlubos @krish2718) to make the final decision.

**Comment by jukkar** [comment]
I revisited and refactored this according to latest comments. The feature is only available for Ethernet based interfaces, also there is no runtime configuration so all Ethernet interfaces will honor the config option `CONFIG_NET_L2_ETHERNET_RESERVE_HEADER` for this.

**Review by pdgendt — APPROVED** [review]

File: subsys/net/ip/net_pkt.c [hunk 15]
```diff
@@ -1177,26 +1199,92 @@ int net_pkt_alloc_buffer(struct net_pkt *pkt,
...
+
+
+	if (net_pkt_iface(pkt) && net_if_l2(net_pkt_iface(pkt))->alloc != NULL) {
+		ret = net_if_l2(net_pkt_iface(pkt))->alloc(net_pkt_iface(pkt),
+							   pkt, size, proto,
+							   timeout);
+		if (ret != -ENOTSUP) {
+			return ret;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pdgendt) [reply subsys/net/ip/net_pkt.c]: Oops this was a leftover comment from the previous version, I apparently never submitted.

**Review by rlubos — APPROVED** [review]
